torchchat/model_dist.py

```diff
@@ -107,22 +88,38 @@ def setup_caches(self, max_batch_size, max_seq_length):
         )
         self.register_buffer("causal_mask", causal_mask, persistent=True)
 
+    def distribute(self, device_mesh: DeviceMesh):
+        if hasattr(self, "tok_embeddings"):
+            parallelize_module(
+                self.tok_embeddings,
+                device_mesh,
+                RowwiseParallel(input_layouts=Replicate()),
+            )
+        for layer in self.layers.values():
+            layer.distribute(device_mesh)
+        # TODO (kwen2501) : parallelize these
+        """
+        parallelize_module(self.norm, device_mesh, ...)
+        parallelize_module(self.output, device_mesh, ...)
+        """
+
     def forward(self, x: Tensor, input_pos: Optional[Tensor] = None) -> Tensor:
         assert self.freqs_cis is not None, "Caches must be initialized first"
         if input_pos is None:
             input_pos = torch.arange(x.shape[1], device=x.device, dtype=torch.long)
         mask = self.causal_mask[None, None, input_pos]
         freqs_cis = self.freqs_cis[input_pos]
 
-        if self.stage_idx == 0:
-            x: DTensor = self.tok_embeddings(x)
-            # TODO: sequence parallelize this
+        if hasattr(self, "tok_embeddings"):
+            x = self.tok_embeddings(x)
 
         for _, layer in self.layers.items():
             x = layer(x, input_pos, freqs_cis, mask)
 
-        if self.stage_idx == self.n_stages - 1:
+        if hasattr(self, "norm"):
```

Comment:
same as above - wondering if these can be n_stages-1 as before for faster check.